// nuxi/dist/chunks/dev-child.mjs
import { d as defineCommand, c as consola } from '../shared/nuxi.89381a23.mjs';
import 'node:util';
import 'node:path';
import 'node:process';
import 'node:tty';
import { o as overrideEnv } from '../shared/nuxi.2509f57e.mjs';
import { s as sharedArgs, l as legacyRootDirArgs } from '../shared/nuxi.01862219.mjs';
import { c } from '../shared/nuxi.5362f0d4.mjs';
import { createNuxtDevServer } from './dev2.mjs';
import { r as resolve } from '../shared/nuxi.dcd5f40f.mjs';
import 'node:url';
import 'node:events';
import '../shared/nuxi.b5db95f6.mjs';
import 'events';
import 'fs';
import 'path';
import 'util';
import 'stream';
import '../shared/nuxi.47289dfd.mjs';
import '../shared/nuxi.7b65c5e0.mjs';
import '../shared/nuxi.6afad232.mjs';
import 'node:module';
import 'node:fs';
import '../shared/nuxi.ae708cae.mjs';
import 'node:assert';
import 'node:v8';
import 'node:perf_hooks';
import '../shared/nuxi.a685c563.mjs';
import './satisfies.mjs';
import '../shared/nuxi.2155838d.mjs';
import '../shared/nuxi.cc8dd4a9.mjs';
import '../shared/nuxi.5e97164c.mjs';
import 'crypto';
import 'module';
import 'perf_hooks';
import 'os';
import 'vm';
import 'url';
import 'assert';
import 'process';
import 'v8';
import 'tty';
import 'node:stream';
import '../shared/nuxi.46c8cc53.mjs';
import '../shared/nuxi.22ca5717.mjs';
import './index2.mjs';
import 'node:http';
import 'node:https';
import 'node:net';
import 'node:os';
import 'http';
import 'https';
import 'node:child_process';
import 'node:fs/promises';
import '../shared/nuxi.dfc4d6fe.mjs';
import '../shared/nuxi.20a653f0.mjs';

const devChild = defineCommand({
  meta: {
    name: "_dev",
    description: "Run nuxt development server (internal command to start child process)"
  },
  args: {
    ...sharedArgs,
    ...legacyRootDirArgs
  },
  async run(ctx) {
    const logger = consola.withTag("nuxi");
    if (!process.send && !c) {
      logger.warn(
        "`nuxi _dev` is an internal command and should not be used directly. Please use `nuxi dev` instead."
      );
    }
    overrideEnv("development");
    const cwd = resolve(ctx.args.cwd || ctx.args.rootDir || ".");
    const devProxyOptions = JSON.parse(process.env.__NUXT_DEV_PROXY__ || "null") || {};
    const nuxtDev = await createNuxtDevServer({
      cwd,
      overrides: ctx.data?.overrides,
      logLevel: ctx.args.logLevel,
      clear: !!ctx.args.clear,
      dotenv: !!ctx.args.dotenv,
      https: devProxyOptions.https,
      port: process.env._PORT ?? void 0
    });
    function sendIPCMessage(message) {
      if (process.send) {
        process.send(message);
      } else {
        logger.info(
          "Dev server event:",
          Object.entries(message).map((e) => e[0] + "=" + JSON.stringify(e[1])).join(" ")
        );
      }
    }
    nuxtDev.on("loading", (message) => {
      sendIPCMessage({ type: "nuxt:internal:dev:loading", message });
    });
    nuxtDev.on("restart", () => {
      sendIPCMessage({ type: "nuxt:internal:dev:restart" });
    });
    nuxtDev.on("ready", (payload) => {
      sendIPCMessage({ type: "nuxt:internal:dev:ready", port: payload.port });
    });
    await nuxtDev.init();
  }
});

export { devChild as default };
